### PR TITLE
HHH-18724 Take constraint composition type into account when determining nullable state of a Column

### PR DESCRIPTION
**Pull request by wytzevanderploeg**
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-18724

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 1]
```diff
@@ -263,14 +275,40 @@ private static boolean applyConstraints(
...
+			return false;
+		}
+
+		// This check assumes that Hibernate Validator is being used providing an
+		// implementation that gives us the composition type that is being used.
+		try {
+			Method compositionTypeMethod = descriptor.getClass()
+					.getMethod( "getCompositionType" );
+			Object result = compositionTypeMethod.invoke( descriptor );
+			if ( result != null && "or".equalsIgnoreCase( result.toString() ) ) {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:302]: Why `equalsIgnoreCase`?

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 2]
```diff
@@ -263,14 +275,40 @@ private static boolean applyConstraints(
...
+	private static boolean isConstraintCompositionOfTypeOr(ConstraintDescriptor<?> descriptor) {
+		if ( descriptor.getComposingConstraints() == null ||
+				descriptor.getComposingConstraints().isEmpty() ) {
+			return false;
+		}
+
+		// This check assumes that Hibernate Validator is being used providing an
+		// implementation that gives us the composition type that is being used.
+		try {
+			Method compositionTypeMethod = descriptor.getClass()
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:299]: This looks like it's being executed many times.

**Review by gavinking — CHANGES_REQUESTED** [review]
The pull request should target `main`.

Also see my comments.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java (hunk 1 above)
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:302]: No particular reason. Just defensive programming I guess.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java (hunk 2 above)
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:299]: I think I can shortcut it a bit more. Should only need to trigger when there is more than one constraint in the composition.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:318]: I think it should be cached statically.

Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:318]: Ran code coverage on it based on the tests.
For the early-return part this gives 76 hits, 71 return early, 5 fall through.
Let met know if you consider this acceptable. Otherwise some kind of caching needs to be applied.

Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java]: Added static caching.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 3]
```diff
@@ -281,16 +298,43 @@ private static boolean applyConstraints(
...
+				descriptor.getComposingConstraints().size() < 2 ) {
+			return false;
+		}
+
+		return CONSTRAINT_COMPOSITION_TYPE_CACHE.computeIfAbsent( descriptor.getAnnotation().getClass(), unused -> {
+			// This check assumes that Hibernate Validator is being used providing an
+			// implementation that gives us the composition type that is being used.
+			try {
+				Method compositionTypeMethod = descriptor.getClass()
+						.getMethod( "getCompositionType" );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:325]: Wait, is this calling a method of the internal class `ConstraintDescriptorImpl`?

I was expecting this to work be reflecting on `ConstraintComposition` annotations.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 4]
```diff
@@ -328,35 +372,31 @@ private static void applySQLCheck(Column column, String checkConstraint) {
...
-						else {
-							LOG.debugf(
-									"@NotNull was applied to attribute [%s] which is defined (at least partially) " +
-											"by formula(s); formula portions will be skipped",
-									property.getName()
-							);
-						}
+	private static boolean isNotNullDescriptor(ConstraintDescriptor<?> descriptor) {
+		return List.of( NotNull.class, NotEmpty.class, NotBlank.class )
+				.contains( descriptor.getAnnotation().annotationType() );
```
[the remark this answers is not in the document]
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:377]: You are completely right. My bad, will revert.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java (hunk 3 above)
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:325]: Yes, did not think of this tbh. Makes more sense to reflect the annotations present. Will update accordingly.

**Comment by wytzevanderploeg** [comment]
Fixed edge case. Added a few more tests to ensure better test coverage.
Had to change logic a bit to ensure that initial state of the nullable boolean was set based on the first constraint in the collection.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 5]
```diff
@@ -328,35 +414,33 @@ private static void applySQLCheck(Column column, String checkConstraint) {
...
+	}
+
+	private static void markNotNull(Property property) {
+		// single table inheritance should not be forced to null due to shared state
+		if ( !( property.getPersistentClass() instanceof SingleTableSubclass ) ) {
+			//composite should not add not-null on all columns
+			if ( !property.isComposite() ) {
+				for ( Selectable selectable : property.getSelectables() ) {
+					if ( selectable instanceof Column ) {
+						((Column) selectable).setNullable( false );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:431]:
```suggestion
					if ( selectable instanceof Column column ) {
						column.setNullable( false );
```

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 6]
```diff
@@ -257,40 +266,117 @@ private static boolean applyConstraints(
...
+			boolean isNotNull = applyConstraint(
+					descriptor,
+					property,
+					propertyDesc,
+					groups,
+					canApplyNotNull,
+					dialect
+			);
+			hasNotNull = compositionLogic.apply( hasNotNull, isNotNull );
+		}
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:304]: All this looks unnecessarily complicated to me. I don't see why we need to use a function reference here, and I don't see why this can't just be written as a `for` loop.

**Review by gavinking — CHANGES_REQUESTED** [review]

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java (hunk 6 above)
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java]: Fair. Updated.

**Comment by gavinking** [comment]
Thanks, looking good now.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 7]
```diff
@@ -72,6 +75,9 @@ class TypeSafeActivator {
 
 	private static final CoreMessageLogger LOG = Logger.getMessageLogger( MethodHandles.lookup(), CoreMessageLogger.class, TypeSafeActivator.class.getName() );
 
+	private static final Map<Class<? extends Annotation>, Boolean>
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:78]: Storing class references to application classes in a static field is unfortunately going to be a problem, unless you clean up the storage properly. Also, concurrent deployments could potentially run into race conditions, so I would suggest you use a `ThreadLocal<Map<Class<? extends Annotation>, Boolean>>` instead and also make sure you clear it in a finally block within the `activate` method.

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 8]
```diff
@@ -257,12 +264,17 @@ private static boolean applyConstraints(
 			PropertyDescriptor propertyDesc,
 			Set<Class<?>> groups,
 			boolean canApplyNotNull,
+			boolean useOrLogicForComposedConstraint,
 			Dialect dialect) {
-		boolean hasNotNull = false;
+
+		boolean firstItem = true;
+		boolean composedResultHasNotNull = false;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:271]:
```suggestion
		boolean composedResultHasNotNull = !useOrLogicForComposedConstraint;
```

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java [hunk 9]
```diff
@@ -276,19 +288,66 @@ private static boolean applyConstraints(
...
+					hasNotNull |= hasNotNullFromComposingConstraints;
+				}
+			}
 
-				hasNotNull = hasNotNull || hasNotNullFromComposingConstraints;
+			if ( firstItem ) {
+				composedResultHasNotNull = hasNotNull;
+				firstItem = false;
+			}
+			else if ( !useOrLogicForComposedConstraint ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:309]:
```suggestion
			if ( !useOrLogicForComposedConstraint ) {
```

File: hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java (hunk 8 above)
Comment (wytzevanderploeg) [reply hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:271]: This is what I initially had but it leads to an edge case when the first property is nullable in a composed constraint of type OR. (See test case I added for the Address.line3 property) In this case the result should be that the column should be considered nullable (nullable OR not-nullable => nullable) This is partly due to the fact that this function is applied recursively for constraints and I did not feel comfortable enough to completely change this. Thus resulting in this head/tails approach.
